rbd: use go-ceph rbd admin task api for 'trash remove' & 'flatten'

Rakshith-R opened the pull request:
- rebase: use go-ceph v0.13.0
- rbd: use go-ceph rbd admin task api instead of cli   
- rbd: remove redundant rbdVolume.connect() in flattneRbdImage() 
- rbd: remove redundant util.Credentials arg from flattenRbdImage() 
- rbd: refractor deleteImage() & trashRemoveImage() 

Closes: #2186 


New Mgr Not Supported errors thrown by go-ceph
- `API call not implemented server-side: No handler found for 'rbd task add trash remove'` (rook 1.2.0, ceph v:14.2.2)
- `rados: ret=-13, Permission denied: "access denied: does your client key have mgr caps? See http://docs.ceph.com/en/latest/mgr/administrator/#client-authentication"` (when user secret does not have mgr caps)

Rakshith-R committed:
rbd: use go-ceph rbd admin task api instead of cli

This commit adds support to go-ceph rbd task api
`trash remove` and `flatten` instead of using cli
cmds.

Fixes: #2186

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: remove redundant rbdVolume.connect() in flattenRbdImage()

This commit removes `rv.Connect(cr)` since the rbdVolume should
have an active connection in this stage of the function call.

`rv.getCloneDepth(ctx)` will work after a connect to the cluster.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: remove redundant util.Credentials arg from flattenRbdImage()

With introduction of go-ceph rbd admin task api, credentials are
no longer required to be passed as cli cmd is not invoked.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: remove redundant rbdVol.getTrashPath() function

This commit removes rbdVol.getTrashPath() function
since it is no longer being used due to introduction
of go-ceph rbd admin task api for deletion.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: add function (cc *ClusterConnection) GetTaskAdmin()

This function returns new go-ceph TaskAdmin to add
tasks on rbd volumes.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

Rakshith-R commented:
> /retest ci/centos/k8s-e2e-external-storage/1.22

failed due to #2264 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/k8s-e2e-external-storage-1.22/detail/k8s-e2e-external-storage-1.22/531/pipeline

humblec commented on build.env:
```diff
@@ -13,7 +13,7 @@ CSI_IMAGE_VERSION=canary
 # Ceph version to use
 BASE_IMAGE=docker.io/ceph/ceph:v16
-CEPH_VERSION=octopus
+BUILD_TAGS=ceph_preview
```
can we document the ceph_preview thing ?

Rakshith-R replied:
>can we document the ceph_preview thing ?

```suggestion
# ceph_preview tag is required to access go-ceph APIs which are under preview.
BUILD_TAGS=ceph_preview
```

@humblec is this sufficient ?

Madhu-1 replied:
add a link to https://github.com/ceph/go-ceph/blob/master/docs/api-stability.md#preview here

nixpanic replied:
I do not think we should set `ceph_preview` as default. Make sure that APIs from go-ceph are stable and part of the latest release.

Rakshith-R replied:
`ceph_preview` is removed and go-ceph is upgraded to v0.13.0 in this pr instead

Madhu-1 commented on internal/util/connection.go:
```diff
@@ -123,6 +123,16 @@ func (cc *ClusterConnection) GetRBDAdmin() (*ra.RBDAdmin, error) {
 	return ra.NewFromConn(cc.conn), nil
 }
 
+// GetTaskAdmin returns TaskAdmin to add tasks on rbd volumes.
```
`rbd volumes` to `rbd images`

Madhu-1 reviewed (commented):
LGTM . @Rakshith-R have you verified its working with older ceph releases where we don't have manager command support?

Rakshith-R commented:
> LGTM . @Rakshith-R have you verified its working with older ceph releases where we don't have manager command support?

Yes, I ran it with ceph 14.2.2 where there's rbd task support, works as expected.

Madhu-1 commented:
> > LGTM . @Rakshith-R have you verified its working with older ceph releases where we don't have manager command support?
> 
> Yes, I ran it with ceph 14.2.2 where there's rbd task support, works as expected.

you mean no rbd task support?

Rakshith-R commented:
> > > LGTM . @Rakshith-R have you verified its working with older ceph releases where we don't have manager command support?
> > 
> > 
> > Yes, I ran it with ceph 14.2.2 where there's rbd task support, works as expected.
> 
> you mean no rbd task support?

yes :sweat_smile:  
*ceph 14.2.2 where there's * *no* * rbd task support, works as expected.

nixpanic reviewed (changes requested):
Suggestion to squash the last two patches together (drop `cr` from `flattenRbdImage()`).

Having `ceph_preview` as default in the `BUILD_TAGS` is not very clean, we should probably wait until go-ceph provides a version where there is no need to set the build tag.

Rakshith-R committed:
rebase: update go-ceph to release v0.13.0

This commit update go-ceph to v0.13.0 and
change CEPH_VERSION in build.env to pacific.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
Resolved/ satisfied golangci linter, ptal thanks

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -700,14 +683,17 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+	log.DebugLog(ctx, "rbd: adding task to remove image %s with id %s from trash", pOpts, pOpts.ImageID)
+
+	ta := ra.Task()
+	_, err = ta.AddTrashRemove(admin.NewImageSpec(pOpts.Pool, pOpts.RadosNamespace, pOpts.ImageID))
```
Please introduce a helper to get the `ImageSpec` from `rbdImage`. There should be only a single location where `admin.NewImageSpec()` is called.

```go
func (ri *rbdImage) getImageSpec() *admin.NewImageSpec {
    return admin.NewImageSpec(ri.Pool, ri.RadosNamespace, ri.ImageID)
}
```

Rakshith-R replied:
> Please introduce a helper to get the `ImageSpec` from `rbdImage`. There should be only a single location where `admin.NewImageSpec()` is called.
> 
> ```go
> func (ri *rbdImage) getImageSpec() *admin.NewImageSpec {
>     return admin.NewImageSpec(ri.Pool, ri.RadosNamespace, ri.ImageID)
> }
> ```

Actually task required imageID and flatten requires imageName, at go-ceph level it was decided to have single `imageSpec` but for ceph it is `imageIDSpec` and `imageSpec`. 

We would still need two methods, one for id and one for name. 
I prefer to keep it the same as it is now.

nixpanic replied:
ah

> We would still need two methods, one for id and one for name.

That is a little sad. The current approach is hard to read, and (as you see with my comment). What is the reason for passing different parameters to `admin.NewImageSpec()`? In case one of the parameters is not set, can a helper function not detect that?

Rakshith-R replied:
> ah
> 
> > We would still need two methods, one for id and one for name.
> 
> That is a little sad. The current approach is hard to read, and (as you see with my comment). What is the reason for passing different parameters to `admin.NewImageSpec()`?

To avoid two almost similar types at go-ceph level 😂

> In case one of the parameters is not set, can a helper function not detect that?

That would be moving the logic inside a function with a boot parameter.

There's no way to automatically detect it. 
If another level of abstraction with helper function is needed, I'll add it but anyway we would have to decide whether it is imageid(trash ops) or imagename.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -700,14 +683,17 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
 	}
-	rbdCephMgrSupported, err := addRbdManagerTask(ctx, pOpts, args)
+
+	log.DebugLog(ctx, "rbd: adding task to remove image %s with id %s from trash", pOpts, pOpts.ImageID)
```
I am a big fan of using `%q` for quoted values in logs, it increases readability a little

nixpanic reviewed (changes requested):
just a nit to improve readability and maintenance of this change, looks good otherwise.

Rakshith-R committed:
cleanup: use %q instead of %s for logging

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
@mergifyio rebase

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -651,74 +624,77 @@ func (rv *rbdVolume) ensureImageCleanup(ctx context.Context) error {
 // deleteImage deletes a ceph image with provision and volume options.
-func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) error {
-	image := pOpts.RbdImageName
+func (rv *rbdVolume) deleteImage(ctx context.Context) error {
```
make this a method of `rbdImage` struct?

nixpanic replied:
that is reported as #2742

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -651,74 +624,77 @@ func (rv *rbdVolume) ensureImageCleanup(ctx context.Context) error {
 // trashRemoveImage adds a task to trash remove an image using ceph manager if supported,
 // otherwise removes the image from trash.
-func trashRemoveImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) error {
+func (rv *rbdVolume) trashRemoveImage(ctx context.Context) error {
```
same here make this method of `rbdImage` struct?

nixpanic replied:
Still pending? if there is a reason not to do that, document it.

Rakshith-R replied:
Opened issue to track it https://github.com/ceph/ceph-csi/issues/2742

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
@Madhu-1 , moving those two methods to rbdImage struct would require moving a few more methods(getImageID, ensureTrashCleanup ...) onto rbdImage struct. 

I would prefer it be done in another pr, maybe we can decide on which methods should be on rbdImage and rbdVolume struct through a issue maybe before making the changes?

@ceph/ceph-csi-contributors PTAL

Rakshith-R commented:
@mergifyio rebase

Madhu-1 commented:
@humblec you have requested changes on this PR. PTAL

humblec commented on internal/rbd/migration.go:
```diff
@@ -112,6 +112,14 @@ func genVolFromMigVolID(ctx context.Context, migVolID *migrationVolID, cr *util.
 		return nil, err
 	}
+	defer rv.Destroy()
+	// if connected , delete it
```
please remove the extra space.

nixpanic replied:
ok!

humblec commented on internal/rbd/migration.go:
```diff
@@ -112,6 +112,14 @@ func genVolFromMigVolID(ctx context.Context, migVolID *migrationVolID, cr *util.
+	// if connected , delete it
+	err = rv.deleteImage(ctx)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to delete rbd image : %s, err: %v", rv.RbdImageName, err)
```
rbd image:

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -828,14 +826,13 @@ func (rv *rbdVolume) flattenRbdImage(
 	}
 
-	ra, err := rv.conn.GetRBDAdmin()
+	log.DebugLog(ctx, "rbd: adding task to flatten image %s", rv)
```
image: ?

nixpanic replied:
using `%q` instead, `:` is usually followed by an error

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -871,10 +871,6 @@ func (rv *rbdVolume) flattenRbdImage(
 			"task manager does not support flatten,image will be flattened once hardlimit is reached: %v",
 			err)
 		if forceFlatten || depth >= hardlimit {
-			err = rv.Connect(cr)
```
The commit has a typo on `flattne`

Rakshith-R commented:
few lines creeped in to migration.go during rebase, fixed that. 
PTAL

humblec commented:
> few lines creeped in to migration.go during rebase, fixed that. PTAL

Thanks @Rakshith-R , I will let @nixpanic to confirm once, as I heard previous changes were tested too.
